src/lib.rs

```diff
@@ -171,108 +170,146 @@ use std::sync::atomic::{AtomicIsize, Ordering, ATOMIC_ISIZE_INIT};
 
 use apm::Listener;
 use bson::Bson;
-use common::{ReadPreference, ReadMode, WriteConcern};
+use common::{ReadPreference, WriteConcern};
 use connstring::ConnectionString;
 use db::{Database, ThreadedDatabase};
 use error::Error::ResponseError;
 use pool::PooledStream;
-use stream::StreamConnector;
-use topology::{Topology, TopologyDescription, TopologyType, DEFAULT_HEARTBEAT_FREQUENCY_MS,
-DEFAULT_LOCAL_THRESHOLD_MS, DEFAULT_SERVER_SELECTION_TIMEOUT_MS};
+use stream::ConnectMethod;
+use topology::{Topology, TopologyDescription, DEFAULT_HEARTBEAT_FREQUENCY_MS,
+               DEFAULT_LOCAL_THRESHOLD_MS, DEFAULT_SERVER_SELECTION_TIMEOUT_MS};
 use topology::server::Server;
 
 /// Interfaces with a MongoDB server or replica set.
 pub struct ClientInner {
-    /// Indicates how a server should be selected for read operations.
-    pub read_preference: ReadPreference,
-    /// Describes the guarantees provided by MongoDB when reporting the success of a write
-    /// operation.
-    pub write_concern: WriteConcern,
     req_id: Arc<AtomicIsize>,
     topology: Topology,
     listener: Listener,
     log_file: Option<Mutex<File>>,
 }
 
 /// Configuration options for a client.
-#[derive(Default)]
-pub struct ClientOptions {
+pub struct Connector {
     /// File path for command logging.
     pub log_file: Option<String>,
-    /// Client-level server selection preferences for read operations.
-    pub read_preference: Option<ReadPreference>,
-    /// Client-level write guarantees when reporting a write success.
-    pub write_concern: Option<WriteConcern>,
     /// Frequency of server monitor updates; default 10000 ms.
     pub heartbeat_frequency_ms: u32,
     /// Timeout for selecting an appropriate server for operations; default 30000 ms.
     pub server_selection_timeout_ms: i64,
     /// The size of the latency window for selecting suitable servers; default 15 ms.
     pub local_threshold_ms: i64,
-    /// Options for how to connect to the server.
-    pub stream_connector: StreamConnector,
 }
 
-impl ClientOptions {
-    /// Creates a new default options struct.
-    pub fn new() -> ClientOptions {
-        ClientOptions {
+impl Default for Connector {
```

Comment:
I realized that deriving default on ClientOptions was a bad idea, as the numerical arguments should not default to 0.